xapian-core/tests/unittest.cc

```diff
@@ -536,11 +536,30 @@ static bool test_tostring1()
     TEST_EQUAL(str(-1), "-1");
     TEST_EQUAL(str(-9), "-9");
     TEST_EQUAL(str(-10), "-10");
+    TEST_EQUAL(str(0x7f), "127");
+    TEST_EQUAL(str(-0x80), "-128");
+    TEST_EQUAL(str(0x7fff), "32767");
+
+    const short min_short = std::numeric_limits<short>::min();
+    std::ostringstream oss;
+    oss << min_short;
+    TEST_EQUAL(str(min_short), oss.str());
+    oss.str("");
+    oss.clear();
+
     TEST_EQUAL(str(0xffffffff), "4294967295");
     TEST_EQUAL(str(0x7fffffff), "2147483647");
     TEST_EQUAL(str(0x7fffffffu), "2147483647");
     TEST_EQUAL(str(-0x7fffffff), "-2147483647");
 
+    const int min_int = std::numeric_limits<int>::min();
+    oss << min_int;
+    TEST_EQUAL(str(min_int), oss.str());
+    oss.str("");
+    oss.clear();
+
+    TEST_EQUAL(str(0x7fffffffffffffff), "9223372036854775807");
+
```

Comment:
On IRC you asked:
> are you happy with the changes i made in my PR. I know we shouldn't use temporary variables but it would get messy if i don't.

My suggestion was to wrap this up in a template function which tests the min and max values for a given type - then the testcase itself can instantiate and call that for each type we want to test - much like the existing testcase here:

```
static bool test_parseunsigned1()
{
    parseunsigned_helper<unsigned char>();
    parseunsigned_helper<unsigned short>();
    parseunsigned_helper<unsigned>();
    parseunsigned_helper<unsigned long>();
    parseunsigned_helper<unsigned long long>();

    return true;
}
```

Then the gory details of the temporary variable and the use of `ostringstream` are encapsulated.

> Also what do you think of using std::to_string instead of ostringstream for testing.? It would result in a much cleaner code and wouldn't need clearing the buffer everytime we use ostringstream.

As I said on IRC `std::to_string()` isn't currently portable enough.  `HACKING` has more details:

> `std::to_string()` - this is completely missing on current versions of mingw and cygwin - in the library, you can `#include "str.h"` and then use the `str()` function instead for most cases.  This is also usually faster than `std::to_string()`.

Of course here we can't use `str()` because we need something to compare the results of `str()` with.

https://gcc.gnu.org/bugzilla/show_bug.cgi?id=52015 seems to be the GCC bug about this.

Reply:
fixed in  a fixup commit.